### DISCUSSION TITLE
form event for before / after saving relations

### DISCUSSION THREAD

**Discussion by ainesophaur** [filamentphp/filament#2801] [Help]
Is there a way to receive an event from filament forms when the form is starting to save any relationships and then after its completed?

**Comment by danharrin** [comment]
What's the goal?

**Comment by ainesophaur** [comment]
The main goal is to use the [batch logging](https://spatie.be/docs/laravel-activitylog/v4/advanced-usage/batch-logs) feature of `spatie/laravel-activitylog`. When saving relationships, it creates single entries to the activity log for each relationship change. By having an event that I can hook into, I can enable batch mode to log all model updates and relationship changes to a single activity log.

**Reply by danharrin** [reply]
Sounds like you should probably use a model observer then? That would be the most reliable solution.

**Reply by ainesophaur** [reply]
When I originally filed this issue I wasn't as comfortable/familiar with the inner workings/magic behind `filament/forms`. Now looking back, I'm not sure my original request of "knowing when relationships are being saved" is needed since its inherently already there. Correct me if I'm wrong, but the logic for relationship updating only happens by calling `$form->getState()` or `$form->getState(true)`... 

If for some reason I needed to independently know the "relationship save" point vs update, the below would be true, correct?
```php
// some class
function submit() {
  //relationship saving
  $data = $this-form->getState();
  //relationship saved
  //model updating
  $this->model->update($data);
  //model updated
}
```

**Reply by danharrin** [reply]
Yeah that is correct for edit forms, but not for create forms. Because the parent model has not been created yet, so the relationships cant be saved. In that case, the relationships are saved after the model has been.

**Comment by studiowizjo** [comment]
You can always create your own `create()` method for creation or `save()` for updating. In this methods there are called lifecycle hooks and you can place your own. Check **CreateRecord** and **EditRecord** classes for code to steal. Because when creating, relationships are saved just after creation (calling method `handleRecordCreation()`), maybe it could be better to overwrite only this smaller method and place your hook at the end of this.
